### DISCUSSION TITLE
how to use `el-dialog` as a child component

### DISCUSSION THREAD

**Discussion by premdasvm** [element-plus/element-plus#6429] [Q&A]
hey, I'm using Element UI with composition API, I've compoent as follows,

Now when i press the button on parent component my modal does show up, but when i close the modal, the value of `leadFormModal` inside parent compoent styas true, which preventing me from opening the modal again unless i refresh, how do i fix this?  i can close the  dialog from it's built in close button if i use v-model instead of my custom props, but using v-model doesn't really allow me to close the modal programmatically, i need to clos ethe modal once a function finish running, like after the save is successful.

`ChildComponent.vue`
```jsx
<template>
  <el-dialog center :model-value="leadFormDialog">
    .....
    <button @click="save" >save</button>
  </el-dialog>
</template>

<script lang="ts">
import { defineComponent, ref } from "vue";
export default defineComponent({
  name: "Child",
  props: {
    leadFormDialog: { type: Boolean, required: true },
  },
  setup(props, { emit }) {
      function save(){
          .......
          ........
          emit('update:leadFormDialog', false)
      }
    };
    return {
      save
    };
  },
});
</script>
```

`ParentComponent.vue`
```html

<template>
<button @click="openModal"></button>
........
 <LeadFormDialog :leadFormDialog="leadFormModal" />
.........
</template>

<script lang="ts">
import { defineComponent, ref } from "vue";
export default defineComponent({
  name: "Parent",
  setup() {
    const leadFormModal = ref(false);

    function openModal(){
      leadFormModal.value = true
    }

     
    return {
      save,
      leadFormModal
    };
  },
});
</script>
```

**Comment by purepear** [comment]
Hey, take a look at these almost identical implementations:

[Element playground 1 (`:before-close` prop)](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtYnV0dG9uIEBjbGljaz1cIm9wZW5EaWFsb2dcIj5PcGVuPC9lbC1idXR0b24+XG4gIDxEaWFsb2cgdi1tb2RlbDp2aXNpYmxlPVwiZGlhbG9nVmlzaWJsZVwiIC8+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IHNldHVwPlxuaW1wb3J0IHsgcmVmIH0gZnJvbSBcInZ1ZVwiXG5pbXBvcnQgRGlhbG9nIGZyb20gJy4vRGlhbG9nLnZ1ZSdcblxuY29uc3QgZGlhbG9nVmlzaWJsZSA9IHJlZihmYWxzZSlcblxuZnVuY3Rpb24gb3BlbkRpYWxvZygpe1xuICBkaWFsb2dWaXNpYmxlLnZhbHVlID0gdHJ1ZVxufVxuPC9zY3JpcHQ+IiwiaW1wb3J0LW1hcC5qc29uIjoie1wiaW1wb3J0c1wiOnt9fSIsIkRpYWxvZy52dWUiOiI8dGVtcGxhdGU+XG4gIDxlbC1kaWFsb2cgOm1vZGVsLXZhbHVlPVwicHJvcHMudmlzaWJsZVwiIDpiZWZvcmUtY2xvc2U9XCJoaWRlXCI+XG4gICAgPGVsLWJ1dHRvbiBAY2xpY2s9XCJzYXZlXCI+c2F2ZTwvZWwtYnV0dG9uPlxuICA8L2VsLWRpYWxvZz5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tIFwidnVlXCI7XG4gIFxuY29uc3QgcHJvcHMgPSBkZWZpbmVQcm9wcyh7XG4gIHZpc2libGU6IHsgdHlwZTogQm9vbGVhbiwgZGVmYXVsdDogZmFsc2UgfVxufSlcblxuY29uc3QgZW1pdCA9IGRlZmluZUVtaXRzKFsndXBkYXRlOnZpc2libGUnXSlcblxuZnVuY3Rpb24gc2F2ZSgpe1xuICAvLyBkbyBzdGhcbiAgaGlkZSgpXG59XG5mdW5jdGlvbiBoaWRlKCkge1xuICBlbWl0KCd1cGRhdGU6dmlzaWJsZScsIGZhbHNlKVxufVxuPC9zY3JpcHQ+In0=)
[Element playground 2 (`@update:modelValue` event)](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuPGVsLWJ1dHRvbiBAY2xpY2s9XCJvcGVuRGlhbG9nXCI+T3BlbjwvZWwtYnV0dG9uPlxuPERpYWxvZyB2LW1vZGVsOnZpc2libGU9XCJkaWFsb2dWaXNpYmxlXCIgLz5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tIFwidnVlXCJcbmltcG9ydCBEaWFsb2cgZnJvbSAnLi9EaWFsb2cudnVlJ1xuXG5jb25zdCBkaWFsb2dWaXNpYmxlID0gcmVmKGZhbHNlKVxuXG5mdW5jdGlvbiBvcGVuRGlhbG9nKCl7XG4gIGRpYWxvZ1Zpc2libGUudmFsdWUgPSB0cnVlXG59XG48L3NjcmlwdD4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XCJpbXBvcnRzXCI6e319IiwiRGlhbG9nLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGVsLWRpYWxvZyA6bW9kZWwtdmFsdWU9XCJwcm9wcy52aXNpYmxlXCIgQHVwZGF0ZTptb2RlbFZhbHVlPVwiZW1pdCgndXBkYXRlOnZpc2libGUnLCAkZXZlbnQpXCI+XG4gICAgPGVsLWJ1dHRvbiBAY2xpY2s9XCJzYXZlXCI+c2F2ZTwvZWwtYnV0dG9uPlxuICA8L2VsLWRpYWxvZz5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tIFwidnVlXCI7XG4gIFxuY29uc3QgcHJvcHMgPSBkZWZpbmVQcm9wcyh7XG4gIHZpc2libGU6IHsgdHlwZTogQm9vbGVhbiwgZGVmYXVsdDogZmFsc2UgfVxufSlcblxuY29uc3QgZW1pdCA9IGRlZmluZUVtaXRzKFsndXBkYXRlOnZpc2libGUnXSlcblxuZnVuY3Rpb24gc2F2ZSgpe1xuICAvLyBkbyBzdGhcbiAgaGlkZSgpXG59XG5mdW5jdGlvbiBoaWRlKCkge1xuICBlbWl0KCd1cGRhdGU6dmlzaWJsZScsIGZhbHNlKVxufVxuPC9zY3JpcHQ+In0=)

The examples use `<script setup>` but you should get the idea.

Learning how v-model works is important here:
* [v-model](https://vuejs.org/guide/components/events.html#usage-with-v-model) - el-dialog uses it
* [v-model:visible](https://vuejs.org/guide/components/events.html#v-model-arguments) - we used `v-model:visible` but we could also use the standard `v-model` and rename `visible` to `modelValue` instead.

**Reply by zeshhaan** [reply]
@purepear Thanks for answering! I have been playing around with it for quite some time and stuck with a similar problem. I can get it working if i assign a `v-model` with [argument](https://vuejs.org/guide/components/events.html#v-model-arguments). But if i use standard `v-model`, it doesn't work. Do you know what might be missing here?

Here is the issue reproduced 
https://stackblitz.com/edit/vitejs-vite-ei2upe?file=src/components/DatePicker.vue

**Reply by purepear** [reply]
Hi @zeshhaan, looks like your issue is not related to el-dialog but changing DatePicker.vue to this should work (if i understand your problem correctly):
```vue
<template>
  <el-date-picker
    :model-value="modelValue"
    @update:modelValue="$emit('update:modelValue', $event)"
  >
  </el-date-picker>
</template>

<script setup>
import { ref } from 'vue';

defineProps(['modelValue']);
defineEmits(['update:modelValue']);
</script>
```

**Reply by zeshhaan** [reply]
Oh yeah, that worked! Adding the modifier to `@update` was the key 💡Thanks so much for the help @purepear !!
